eppo_core/src/str.rs

```diff
@@ -97,6 +97,12 @@ impl std::ops::Deref for Str {
     }
 }
 
+impl std::borrow::Borrow<str> for Str {
+    fn borrow(&self) -> &str {
+        self.as_str()
+    }
+}
```

Comment:
This allows us using `&str` for querying `HashMap<Str, _>`.

Quite a bit of changes in this PR is changing `String` -> `Str` to enable quick cloning.

I start leaning to using `Str` whenever we use strings because it's much cheaper to clone and has small string optimization (small strings are stored without heap allocation), so it's a good fit for our use case. These optimizations also make it much faster for converting to/from other languages (e.g., Python). The only downside is that occasional `String -> Str` conversion might be a bit more costly but there are few